### PR TITLE
feat(ui_firestore): use aggregate query to display total rows count in DataTable

### LINKED ISSUE #9745: 🐛 [ui_firestore] Facilitate the .count() capabilities

### PR DESCRIPTION
**Pull request by lesnitsky**
## Description

See title

## Related Issues

Closes #9745

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/firebase/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Review by Lyokone — APPROVED** [review]

File: packages/firebase_ui_firestore/lib/src/query_builder.dart [hunk 1]
```diff
@@ -502,3 +502,48 @@ class FirestoreListView<Document> extends FirestoreQueryBuilder<Document> {
...
+  State<AggregateQueryBuilder> createState() => _AggregateQueryBuilderState();
+}
+
+class _AggregateQueryBuilderState extends State<AggregateQueryBuilder> {
+  late var queryFuture = widget.query.get();
+
+  @override
+  Widget build(BuildContext context) {
+    return FutureBuilder<AggregateQuerySnapshot>(
+      key: ValueKey(queryFuture),
```
Comment (rrousselGit) [inline packages/firebase_ui_firestore/lib/src/query_builder.dart:533]: Why is this here?

File: packages/firebase_ui_firestore/lib/src/query_builder.dart [hunk 2]
```diff
@@ -502,3 +502,48 @@ class FirestoreListView<Document> extends FirestoreQueryBuilder<Document> {
...
+    required this.query,
+    required this.builder,
+  });
+
+  @override
+  State<AggregateQueryBuilder> createState() => _AggregateQueryBuilderState();
+}
+
+class _AggregateQueryBuilderState extends State<AggregateQueryBuilder> {
+  late var queryFuture = widget.query.get();
```
Comment (rrousselGit) [inline packages/firebase_ui_firestore/lib/src/query_builder.dart:528]: There's the possible case where `widget.query` changes

Comment (rrousselGit) [reply packages/firebase_ui_firestore/lib/src/query_builder.dart:528]: Oops ignore me

File: packages/firebase_ui_firestore/lib/src/query_builder.dart [hunk 3]
```diff
@@ -502,3 +502,48 @@ class FirestoreListView<Document> extends FirestoreQueryBuilder<Document> {
...
+    );
+  }
+
+  @override
+  void didUpdateWidget(covariant AggregateQueryBuilder oldWidget) {
+    if (widget.query != oldWidget.query) {
+      queryFuture = widget.query.get();
+    }
+
+    super.didUpdateWidget(oldWidget);
```
Comment (rrousselGit) [inline packages/firebase_ui_firestore/lib/src/query_builder.dart:547]:
```suggestion
    super.didUpdateWidget(oldWidget);
    if (widget.query != oldWidget.query) {
      queryFuture = widget.query.get();
    }

```

Comment (rrousselGit) [reply packages/firebase_ui_firestore/lib/src/query_builder.dart:547]: Wow, git didn't like that.
I just moved the super above

File: packages/firebase_ui_firestore/lib/src/table_builder.dart [hunk 4]
```diff
@@ -836,12 +849,16 @@ class _Source extends DataTableSource {
   @override
   int get selectedRowCount => _selectedRowIds.length;
 
+  int? _aggregateCount;
```
Comment (rrousselGit) [inline packages/firebase_ui_firestore/lib/src/table_builder.dart:852]: I'd store the snapshot itself, like with `_previousSnapshot`

File: packages/firebase_ui_firestore/lib/src/table_builder.dart [hunk 5]
```diff
@@ -836,12 +849,16 @@ class _Source extends DataTableSource {
...
+
   @override
   bool get isRowCountApproximate =>
-      _previousSnapshot!.isFetching || _previousSnapshot!.hasMore;
+      _aggregateCount != null ||
```
Comment (rrousselGit) [inline packages/firebase_ui_firestore/lib/src/table_builder.dart:856]: Shouldn't this be `== null`?

File: packages/firebase_ui_firestore/lib/src/query_builder.dart (hunk 1 above)
Comment (lesnitsky) [reply packages/firebase_ui_firestore/lib/src/query_builder.dart:533]: to get rid of redundant FutureBuilder rebuilds on hot reload (reassemble)

Comment (rrousselGit) [reply packages/firebase_ui_firestore/lib/src/query_builder.dart:533]: This has the nasty effect of destroying the entire subtree whenever the query changes

I don't quite understand which redundant rebuild you're referring to.
But this possibly negatively impacts production code

**Review by rrousselGit — APPROVED** [review]
